optuna/storages/redis_storage.py

```diff
@@ -0,0 +1,511 @@
+        return distribution.to_internal_repr(self.get_trial(trial_id).params[param_name])
+
+    def set_trial_value(self, trial_id, value):
+        # type: (int, float) -> None
+
+        with self._lock:
+            trial = self.get_trial(trial_id)
+            self.check_trial_is_updatable(trial_id, trial.state)
+
+            trial.value = value
+            self.redis.set("trial_id:{:010d}:frozentrial".format(trial_id), pickle.dumps(trial))
+
+    def _update_cache(self, trial_id):
+        # type: (int) -> None
+
+        trial = self.get_trial(trial_id)
+        if trial.state != structs.TrialState.COMPLETE:
+            return
+        study_id = self.get_study_id_from_trial_id(trial_id)
+        if not self.redis.exists("study_id:{:010d}:best_trial_id".format(study_id)):
+            self.set_best_trial(study_id, trial_id)
+            return
+
+        best_value = float(self.get_best_trial(study_id).value)
+        new_value = float(trial.value)
+
+        # Complete trials do not have `None` values.
+        assert new_value is not None
+
+        if self.get_study_direction(study_id) == structs.StudyDirection.MAXIMIZE:
+            if new_value > best_value:
+                self.set_best_trial(study_id, trial_id)
+        else:
+            if new_value < best_value:
+                self.set_best_trial(study_id, trial_id)
+
+        return
+
+    def set_trial_intermediate_value(self, trial_id, step, intermediate_value):
+        # type: (int, int, float) -> bool
+
+        with self._lock:
+            self.check_trial_is_updatable(trial_id, self.get_trial(trial_id).state)
+
+            frozen_trial = self.get_trial(trial_id)
+            if step in frozen_trial.intermediate_values:
+                return False
+
+            frozen_trial.intermediate_values[step] = intermediate_value
+            self.set_trial(trial_id, frozen_trial)
+
+            return True
+
+    def set_trial_user_attr(self, trial_id, key, value):
+        # type: (int, str, Any) -> None
+
+        with self._lock:
+            trial = self.get_trial(trial_id)
+            self.check_trial_is_updatable(trial_id, trial.state)
+            trial.user_attrs[key] = value
+            self.set_trial(trial_id, trial)
+
+    def set_trial_system_attr(self, trial_id, key, value):
+        # type: (int, str, Any) -> None
+
+        with self._lock:
+            trial = self.get_trial(trial_id)
+            self.check_trial_is_updatable(trial_id, trial.state)
+            trial.system_attrs[key] = value
+            self.set_trial(trial_id, trial)
+
+    def get_trial(self, trial_id):
+        # type: (int) -> structs.FrozenTrial
+
+        return pickle.loads(self.redis.get("trial_id:{:010d}:frozentrial".format(trial_id)))
+
+    def set_trial(self, trial_id, trial):
+        # type: (int, structs.FrozenTrial) -> None
+
+        self.redis.set("trial_id:{:010d}:frozentrial".format(trial_id), pickle.dumps(trial))
+
+    def del_trial(self, trial_id):
```

Comment:
How about prefixing this method with an underscore making it obviously private?

Reply:
Right. Prefixed all private methods with _ make the intention clearer. DONE